File: solutions/2026-01-25/cpp/2026-01-25-minimum-difference-between-highest-and-lowest-of-k-scores.cpp
// Problem: Minimum Difference Between Highest and Lowest of K Scores
// Link: https://leetcode.com/problems/minimum-difference-between-highest-and-lowest-of-k-scores/
// Approach: The problem asks us to find the minimum difference between the highest and lowest scores among any k selected students.
// To minimize this difference, we should select k students whose scores are as close to each other as possible.
// This suggests sorting the array of scores first. Once the array is sorted, any contiguous subarray of length k will represent a set of k scores.
// The difference between the highest and lowest score in such a subarray will be the last element minus the first element.
// We can iterate through all possible contiguous subarrays of length k and calculate this difference, keeping track of the minimum difference found.
// This is a classic sliding window approach on a sorted array. We maintain a window of size k.
//
// Time Complexity: O(N log N) due to sorting the array, where N is the number of students (nums.length).
// The sliding window part takes O(N) time.
// Space Complexity: O(log N) or O(N) depending on the sorting algorithm used by the standard library.
// If in-place sorting is used, it's O(log N) for recursion stack. If a copy is made, it's O(N).

#include <vector>
#include <algorithm>
#include <climits>

class Solution {
public:
    int minimumDifference(std::vector<int>& nums, int k) {
        // If k is 1, the difference is always 0 as we pick only one score.
        if (k == 1) {
            return 0;
        }

        // Sort the array of scores in ascending order.
        // This is crucial because to minimize the difference between the highest and lowest score within a group of k,
        // those k scores should be adjacent in the sorted list.
        std::sort(nums.begin(), nums.end());

        // Initialize the minimum difference to the maximum possible integer value.
        // This variable will store the smallest difference found so far.
        int min_diff = INT_MAX;

        // Iterate through the sorted array using a sliding window of size k.
        // The window starts at index `i` and ends at index `i + k - 1`.
        // The loop runs from `i = 0` up to `nums.size() - k` to ensure the window of size k is always within bounds.
        for (int i = 0; i <= (int)nums.size() - k; ++i) {
            // For the current window, the lowest score is at index `i` (nums[i])
            // and the highest score is at index `i + k - 1` (nums[i + k - 1]).
            // Calculate the difference between the highest and lowest scores in this window.
            int current_diff = nums[i + k - 1] - nums[i];

            // Update `min_diff` if the `current_diff` is smaller than the current `min_diff`.
            min_diff = std::min(min_diff, current_diff);
        }

        // Return the minimum difference found among all possible groups of k scores.
        return min_diff;
    }
};

File: solutions/2026-01-25/java/2026-01-25-minimum-difference-between-highest-and-lowest-of-k-scores.java
```java
import java.util.Arrays;

/**
 * Problem: Minimum Difference Between Highest and Lowest of K Scores
 * LeetCode Link: https://leetcode.com/problems/minimum-difference-between-highest-and-lowest-of-k-scores/
 *
 * Approach:
 * To minimize the difference between the highest and lowest scores of k students,
 * we should select k scores that are as close to each other as possible.
 * Sorting the array allows us to easily consider contiguous subarrays of size k.
 * Once the array is sorted, we can use a sliding window of size k.
 * For each window, the difference between the highest and lowest score is simply
 * the difference between the last element and the first element of that window.
 * We iterate through all possible windows and keep track of the minimum difference found.
 *
 * Time Complexity:
 * O(N log N) due to the sorting step, where N is the number of students (nums.length).
 * The sliding window part takes O(N) time.
 *
 * Space Complexity:
 * O(1) if the sorting is done in-place. If a copy of the array is made for sorting,
 * then it would be O(N). The problem constraints allow in-place sorting.
 */
class Solution {
    public int minimumDifference(int[] nums, int k) {
        // If k is 1, the difference is always 0 as we pick only one score.
        if (k == 1) {
            return 0;
        }

        // Sort the array of scores in ascending order.
        // This is crucial for the sliding window approach to find contiguous elements.
        Arrays.sort(nums);

        // Initialize the minimum difference to the maximum possible value.
        // This will be updated as we find smaller differences.
        int minDiff = Integer.MAX_VALUE;

        // Use a sliding window of size k.
        // The window starts at index `i` and ends at index `i + k - 1`.
        // The loop runs until `i + k - 1` is within the bounds of the array.
        for (int i = 0; i <= nums.length - k; i++) {
            // For the current window, the lowest score is nums[i] and the highest is nums[i + k - 1].
            // Calculate the difference between the highest and lowest score in this window.
            int currentDiff = nums[i + k - 1] - nums[i];

            // Update minDiff if the current difference is smaller.
            minDiff = Math.min(minDiff, currentDiff);
        }

        // Return the minimum difference found across all windows.
        return minDiff;
    }
}
```